File: web/python/examples/live_semantic_scene.py
"""Paired Python proof for typed live membership and property publication.

Run in the Noon browser/Pyodide authoring environment.  This calls the real
canonical Rust/WASM context and its live execution player; it does not encode a
scene document or synchronize a Python state copy.
"""

from noon import Circle, Scene, Square


class LiveSemanticScene(Scene):
    def construct(self):
        anchor = Circle(0.5)
        toggled = Circle(1.0)
        appended = Square(1.5)
        self.add(anchor)
        self.add(toggled)

        live = self.live_execution()
        live.remove(toggled)
        live.add(toggled)
        live.add(appended)
        live.set_translation(appended, 2.0, -1.0)

        assert live.effective_center(anchor) == (0.0, 0.0)
        assert live.effective_center(appended) == (2.0, -1.0)

File: crates/noon-native/examples/live_semantic_scene.rs
//! Paired Rust proof for live property and structural membership publication.
use noon::Scene;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let mut scene = Scene::new();
    let anchor = scene.circle(0.5)?;
    let toggled = scene.circle(1.0)?;
    // Author detached objects before runtime bootstrap so the session starts at the
    // same semantic revision and later publishes only their membership.
    let appended = scene.square(1.5)?;
    scene.add(&anchor)?;
    scene.add(&toggled)?;
    let mut session = scene.execution_session()?;
    let anchor_slot = session.execution_slot_for_frame_index(0).unwrap();

    {
        let mut live = scene.live(&mut session);
        live.remove(&toggled)?;
        assert!(live.effective(&toggled).is_err());
        live.add(&toggled)?;
        live.add(&appended)?;
        live.set_translation(&appended, 2.0, -1.0)?;
        assert_eq!(live.effective(&appended)?.transform.translation.x, 2.0);
    }

    assert_eq!(session.execution_slot_for_frame_index(0), Some(anchor_slot));
    noon_native::run(session)?;
    Ok(())
}
